src/storage/lock_manager/lock_wait_context.rs

```diff
@@ -0,0 +1,254 @@
+//! of a single `AcquirePessimisticLock` request, and ensuring the internal
+//! callback for returning response through RPC is called at most only once.
+//!
+//! Note: The corresponding implementation in `WaiterManager` is not yet
+//! implemented, and this mod is currently not used yet.
+
+use std::{
+    convert::TryInto,
+    result::Result,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+use parking_lot::Mutex;
+use txn_types::TimeStamp;
+
+use crate::storage::{
+    errors::SharedError,
+    lock_manager::{lock_waiting_queue::PessimisticLockKeyCallback, LockManager, LockWaitToken},
+    Error as StorageError, PessimisticLockRes, ProcessResult, StorageCallback,
+};
+
+pub struct LockWaitContextInner {
+    #[allow(dead_code)]
+    pr: ProcessResult,
+    cb: StorageCallback,
+    #[allow(dead_code)]
+    lock_wait_token: LockWaitToken,
+}
+
+pub struct LockWaitContextSharedState {
+    ctx_inner: Mutex<Option<LockWaitContextInner>>,
+    pub finished: AtomicBool,
+}
+
+impl LockWaitContextSharedState {
+    pub fn is_finished(&self) -> bool {
+        self.finished.load(Ordering::Acquire)
+    }
+}
+
+#[derive(Clone)]
+pub struct LockWaitContext<L: LockManager> {
+    shared_states: Arc<LockWaitContextSharedState>,
+    #[allow(dead_code)]
+    lock_manager: L,
+    allow_lock_with_conflict: bool,
+
+    // Fields for logging:
+    start_ts: TimeStamp,
+    for_update_ts: TimeStamp,
+}
+
+impl<L: LockManager> LockWaitContext<L> {
+    pub fn new(
+        lock_manager: L,
+        lock_wait_token: LockWaitToken,
+        start_ts: TimeStamp,
+        for_update_ts: TimeStamp,
+        first_batch_pr: ProcessResult,
+        cb: StorageCallback,
+        allow_lock_with_conflict: bool,
+    ) -> Self {
+        let inner = LockWaitContextInner {
+            pr: first_batch_pr,
+            cb,
+            lock_wait_token,
+        };
+        Self {
+            shared_states: Arc::new(LockWaitContextSharedState {
+                ctx_inner: Mutex::new(Some(inner)),
+                finished: AtomicBool::new(false),
+            }),
+            lock_manager,
+            allow_lock_with_conflict,
+            start_ts,
+            for_update_ts,
+        }
+    }
+
+    pub fn get_shared_states(&self) -> &Arc<LockWaitContextSharedState> {
+        &self.shared_states
+    }
+
+    /// Get the callback that should be invoked when finishes executing the
+    /// scheduler command that issued the lock-waiting.
+    ///
+    /// When we support partially finishing a pessimistic lock request (i.e.
+    /// when acquiring lock multiple keys in one single request, allowing
+    /// some keys to be locked successfully while the others are blocked or
+    /// failed), this will be useful for handling the result of the first
+    /// write batch. But currently, the first write batch of a lock-waiting
+    /// request is always empty, so the callback is just noop.
```

Comment:
Why is the first write batch of a lock-waiting request always empty?

Reply:
I think it's a placeholder for more complex optimizations in the future. Well, it may be a question whether we should introduce them now.